### PR TITLE
Bigscience176b

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Opening a draft pull request for:

## Testing the BigScience176B tokenizer

To test the tokenizer we:
+ Generate tokens from a dummy text. Encode it and decode it back
+ Make sure that the tokens are the same than the one created by megatron-deepspeed
+ For now (before the tokenizer 0.12.1 release) follow [these instructions](https://huggingface.co/bigscience-catalogue-data-dev/byte-level-bpe-tokenizer-no-norm-250k-whitespace-and-eos-regex-alpha-v3-dedup-lines-articles) to download and use the tokenizer.
+ Reconstruction tests on ```xnli``` dataset

### TO DO:

- [x] Push the .bin data in the Hub to be able to download it in local and use it? OR hardcode the values of the tokens (better)
- [ ] Discuss about slow tests
- [ ] Hardcoding tokens from ```xnli``` dataset

## Testing the embeddings from the converted model

I first:
+ Retrieved the embeddings generated from the GPT2 model that has been trained using Megatron-DeepSpeed. For now the mean value of each embedding is hardcoded.
Then:
+ Load the converted model and compare the values of the mean embeddings using the same input. The check passes **if and only if** the weights of the model class modules are with the same ```dtype``` as the ones computed on the script on Megatron-DeepSpeed. With this implementation should be able to try the test with ```float16```,  ```float32``` or ```bfloat16```and I hardcoded the mean values of the embeddings for each ```dtype```. 

### What do we needed to make it work?

+ Add a specific dtype when loading the module [779bdd0](https://github.com/thomwolf/transformers/pull/1/commits/779bdd09bc3bfbfefcf4cb96baa5a500992ab60d) & [4f17f81](https://github.com/thomwolf/transformers/pull/1/commits/4f17f8151f9c6e6389eb9a9149a8aee95cc805f9)

### TO DO:

The test **passes** for the logits before the layer norm for all supported ```dtype```, but not passing for the logits after the layernorm, I suspect that:
- [x] The weights were not loaded correctly into ```bfloat16``` (exactly [here](https://github.com/thomwolf/transformers/blob/88f05f40100ae9e2f94bacfca5302c2dd77b0ab1/src/transformers/models/bigscience176b/convert_bigscience176b_original_checkpoint_to_pytorch.py#L107), but even after dividing by an ```int```, the weights are still ```bfloat16```)
- [x] We have to implement the ```FusedLayerNorm``` on the model definition, since this is used during training.
- [x] Add more metrics (max, min, median)?
- [x]  Understand why we do not obtain the same values on GPU and CPU

More exciting things to come!

 - [x] Make sure that apex is **not installed** in the virtual env that I am using
 - [x] Check the vales of the attention layer (before the MLP)
 - [x] Figure out how to merge the different outputs across TP ranks -> To discuss with Thomas just to be sure
 - [x] Add biases on each MLP (should work)
 - [x] Check ```dense_4h_to_h```?
 - [x] Check the values of the logits after the MLP
 - [x] Get the final logits

### CODE REVIEW

**Comment by younesbelkada** [comment]
Just added the hardcoded tests for the tokenizer. I will see about slow tests and how to integrate a test with another dataset (eg xnli).

**Comment by younesbelkada** [comment]
Should we keep the ```test_tokenization_bigscience.py``` (which is different from ```test_tokenizer_bigscience.py```)? The first file aims to check some classic operations I guess, as the second one is very specific to BigScience (hardcoded tokens)
cc @thomwolf

File: src/transformers/models/bigscience176b/configuration_bigscience176b.py [hunk 1]
```diff
@@ -114,11 +114,17 @@ def __init__(
...
         initializer_range=0.02,  # TODO
-        use_cache=True,  # TODO
+        use_cache=False,  # TODO
         bos_token_id=0, # 50256,  # TODO
         eos_token_id=0, # =50256,  # TODO
+        apply_residual_connection_post_layernorm=False,
+        bias_dropout_fusion=True,
+        skip_bias_add=True,
+        hidden_dropout=0.1,
+        attention_dropout=0.1,
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/configuration_bigscience176b.py:133]: Can you check all the arguments are in the docstring?

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 2]
```diff
@@ -69,25 +71,48 @@
 _TOKENIZER_FOR_DOC = "BigScience176BTokenizer"
 
 BIGSCIENCE176B_PRETRAINED_MODEL_ARCHIVE_LIST = [
-    "bigscience/BigScience176B",
+    "bigscience/bigscience-small-testing",
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py:66]: Very good that you already added a small model-card

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 3]
```diff
@@ -69,25 +71,48 @@
...
     # See all BigScience176B models at https://huggingface.co/models?filter=bigscience176b
 ]
 
+# Utility functions below:
+
 def attention_mask_func(attention_scores, attention_mask):
         return attention_scores.masked_fill_(attention_mask, -10000.0)
 
+def transpose_tensor(x):
+    return x.transpose(0, 1).contiguous()
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py]: Nitpick: not a huge fan of adding a method for a oneliner

File: src/transformers/models/bigscience176b/scale_mask_softmax.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import enum
+import torch
+import torch.nn as nn
+
+class AttnMaskType(enum.Enum):
+    padding = 1
+    causal = 2
+    prefix = 3
+
+class ScaleMaskSoftmax(nn.Module):
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/scale_mask_softmax.py]: I think we could remove this whole file and replace it by a nn.Softmax in the modeling.py file for a couple of reasons:
- in `transformers` we strongly favor trying to have the whole forward pass in a file file if possible
- keep in mind that this is a pretrained model that will be provided "as is". As such we don't expect people to reuse this class for anything else or to tweak the arguments. In other words can hardcode as much as you want.

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 5]
```diff
@@ -103,19 +128,23 @@ def __init__(self, config, layer_number=None):
...
-        #     attn_mask_type=AttnMaskType.causal,
-        #     scaled_masked_softmax_fusion=config.masked_softmax_fusion,
-        #     mask_func=attention_mask_func,
-        #     softmax_in_fp32=False,
-        #     scale=coeff)
-
-        self.query_key_value = nn.Linear(self.hidden_size, 3 * self.hidden_size)
-        self.dense = nn.Linear(self.hidden_size, self.hidden_size)
+        # self.scale_mask_softmax = nn.Softmax()
+        self.scale_mask_softmax = ScaleMaskSoftmax(  # TODO setup back
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py]: see my comment below on `ScaleMaskSoftmax`

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 6]
```diff
@@ -219,9 +239,17 @@ def forward(
...
 
         # attention scores and attention mask [b, np, sq, sk]
-        attention_probs = self.scale_mask_softmax(attention_scores,
-                                                  attention_mask)
 
+        # softmax across tp ranks
+
+        aggregated_tensors = []
+        slices = self.num_heads / self.pretraining_tp
+        for i in range(self.pretraining_tp):
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py:219]: add a quick comment or link to the pytorch repo issue on why we do this loop

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 7]
```diff
@@ -262,46 +290,83 @@ def forward(
...
 
         outputs = (output, present)
         if output_attentions:
             outputs += (None,) # TODO not implemented yet
 
-        return outputs  # a, present, (attentions)
+        return outputs, output_bias  # a, present, (attentions)
 
 
 # Copied from transformers.models.gpt2.modeling_gpt2.GPT2MLP with GPT2->BigScience176B
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py]: This "Copied from" line is actually a macro that ask the test in the transformers repo to make sure that the class we copied the original model from (`GPT2MLP`) and the class we have now (`BigScience176BMLP`) are identical and otherwise raise an error (you can ask Sylvain Gugger about it). We use it to make sure model stay in sync when they need to across evolution of the repo.

You should probably remove it here since we modified the class and don't expect it to stay identical

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 8]
```diff
@@ -262,46 +290,83 @@ def forward(
...
+                hidden_states,
+                self.dense_h_to_4h.weight
+            ),
+            self.dense_h_to_4h.bias
+        )
+
+        intermediate_output = torch.zeros_like(input_)
+        slices = self.dense_4h_to_h.weight.shape[-1] / self.pretraining_tp
+        for i in range(self.pretraining_tp):
+            intermediate_output = intermediate_output + F.linear(hidden_states[:, :, int(i*slices):int((i+1)*slices)], self.dense_4h_to_h.weight[:, int(i*slices):int((i+1)*slices)])
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py]: make sure you run `make style` or `make fixup` before merging in the main repo (we use code formating tools). Otherwise you'll have an error

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 9]
```diff
@@ -69,25 +71,48 @@
...
+
+
+@torch.jit.script
+def bias_dropout_add_fused_inference(x, bias, residual, prob):
+    # type: (Tensor, Tensor, Tensor, float) -> Tensor
+    return bias_dropout_add(x, bias, residual, prob, False)
+
 class BigScience176BAttention(nn.Module):
-    def __init__(self, config, layer_number=None):
+    def __init__(self, config, layer_number=None, dtype=torch.bfloat16):
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py]: Why not reading the `dtype` from the `config`?

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 10]
```diff
@@ -326,8 +391,35 @@ def get_slopes_power_of_2(n):
         # tp_world_size = mpu.get_tensor_model_parallel_world_size()
         # tp_index = mpu.get_tensor_model_parallel_rank()
         # alibi = alibi.reshape((tp_world_size, -1, *alibi.shape[1:]))[tp_index]
+        alibi = alibi.to(dtype)
 
         return alibi
+    def forward_attention(
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py]: where do we use `forward_attention`?

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 11]
```diff
@@ -569,12 +691,19 @@ def __init__(self, config):
         super().__init__(config)
 
         self.embed_dim = config.hidden_size
+        dtype = getattr(torch, config.dtype)
+        assert dtype in [torch.bfloat16, torch.float32, torch.float, torch.float16]
+
+        self.word_embeddings = nn.Embedding(config.vocab_size, self.embed_dim, dtype=dtype)
+        # self.word_embeddings = nn.Embedding(config.vocab_size, self.embed_dim)
+        # self.word_embeddings_layernorm = MixedFusedLayerNorm(self.embed_dim, dtype=dtype, eps=config.layer_norm_epsilon)
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py]: clean up a bit this init func

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py [hunk 12]
```diff
@@ -680,24 +809,24 @@ def forward(
...
+        #     # used in OpenAI GPT, we just need to prepare the broadcast dimension here.
+        #     attention_mask = attention_mask[:, None, None, :]
+
+        #     # Since attention_mask is 1.0 for positions we want to attend and 0.0 for
+        #     # masked positions, this operation will create a tensor which is 0.0 for
+        #     # positions we want to attend and -10000.0 for masked positions.
+        #     # Since we are adding it to the raw scores before the softmax, this is
+        #     # effectively the same as removing these entirely.
+        #     attention_mask = attention_mask.to(dtype=self.dtype)  # fp16 compatibility
+        #     attention_mask = (1.0 - attention_mask) * -10000.0
```
Comment (thomwolf) [inline src/transformers/models/bigscience176b/modeling_bigscience176b.py:773]: We'll probably need to pass the attention mask at some point but we can see that later when fixing all the integration tests

**Review by thomwolf — APPROVED** [review]
Yes looks pretty good for now.

I've added a couple of comments along the way, you can merge when you are happy with it :)

File: src/transformers/models/bigscience176b/configuration_bigscience176b.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/configuration_bigscience176b.py:133]: Should be resolved [here](https://github.com/thomwolf/transformers/pull/1/commits/6b22a495f20f6d1ead8ac6f1b8d8002d7c7ac4a5)

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 3 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py]: Agree! Should be fixed [here](https://github.com/thomwolf/transformers/pull/1/commits/a74f0f55b6d27df0cd8f901befdacf4dffc9b03b). Made sure also that the tests pass again

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 7 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py]: Perfect ! I was not aware of that macro (which is actually quite useful for some cases I agree!) should be fixed [here](https://github.com/thomwolf/transformers/pull/1/commits/e27e554e15da432d0d0aafb60b554d7c9640bfa9)

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 9 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py]: It is because I assert that the dtype is loaded correctly once [here](https://github.com/younesbelkada/transformers/blob/e27e554e15da432d0d0aafb60b554d7c9640bfa9/src/transformers/models/bigscience176b/modeling_bigscience176b.py#L689), I did that to avoid asserting as many times as you retrieve the dtype from the config. If you think that the assert is not necessary, then I can read the dtypes directly from the config files on all the modules

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 10 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py]: I used to test it during my first tests, removed it [here](https://github.com/younesbelkada/transformers/commit/39644e770ca8cbc8315d1fbce0c19d1e41097c15)

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 11 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py]: Done [here](https://github.com/younesbelkada/transformers/commit/93d270a906a54f3f584d6cbe84196c7088069fa0)!

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 12 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py:773]: Actually I am passing the attention masks in the test (since it is done also on the Megatron-LM script I was using, I wanted to ensure the reproducibility using the exact same input, including the masks). I commented this block because this operation is done automatically inside the sclaed masked softmax function by [this function](https://github.com/younesbelkada/transformers/blob/93d270a906a54f3f584d6cbe84196c7088069fa0/src/transformers/models/bigscience176b/modeling_bigscience176b.py#L81) which is [initialized](https://github.com/younesbelkada/transformers/blob/93d270a906a54f3f584d6cbe84196c7088069fa0/src/transformers/models/bigscience176b/modeling_bigscience176b.py#L134) and used by the scaled sofmax function [here](https://github.com/younesbelkada/transformers/blob/93d270a906a54f3f584d6cbe84196c7088069fa0/src/transformers/models/bigscience176b/scale_mask_softmax.py#L74)

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 6 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py:219]: Fixed [here](https://github.com/younesbelkada/transformers/commit/a01a894a3d92779a000dcba0ca263de888df3c46)

File: src/transformers/models/bigscience176b/scale_mask_softmax.py (hunk 4 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/scale_mask_softmax.py]: Perfect! Thanks for the explanation, fixed everything in [this commit](https://github.com/younesbelkada/transformers/commit/adcfc157b9dabf40cffd0a1b5b685438726b03f8), no need to add an extra file anymore now :100:

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 12 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py:773]: Since the script has changed a bit (Removed Scaled Mask softmax), I adapted everything on the new commit!

File: src/transformers/models/bigscience176b/modeling_bigscience176b.py (hunk 9 above)
Comment (younesbelkada) [reply src/transformers/models/bigscience176b/modeling_bigscience176b.py]: Actually I think that it makes sense to retrieve it directly from the config. One assert at the very beginning should be sufficient. Fixed it [here](https://github.com/younesbelkada/transformers/commit/fc3056d0f024bd3f5ccf76bc7a7483628f1eefbf)

**Comment by thomwolf** [comment]
Ok awesome, great job
